cpp/oneapi/dal/algo/decision_forest/backend/gpu/train_splitter_sp_opt_impl.hpp

```diff
@@ -0,0 +1,121 @@
+    using node_t = node<Index>;
+    using node_list_t = node_list<Index>;
+    using node_group_view_t = node_group_view<Index>;
+
+public:
+    train_splitter_sp_opt_impl() = default;
+    ~train_splitter_sp_opt_impl() = default;
+
+    static sycl::event best_split_single_pass_large(sycl::queue& queue,
+                                                    const context_t& ctx,
+                                                    const pr::ndarray<Bin, 2>& data,
+                                                    const pr::ndview<Float, 1>& response,
+                                                    const pr::ndarray<Index, 1>& tree_order,
+                                                    const pr::ndarray<Index, 1>& selected_ftr_list,
+                                                    const pr::ndarray<Index, 1>& bin_offset_list,
+                                                    const imp_data_t& imp_data_list,
+                                                    const pr::ndarray<Index, 1>& node_ind_list,
+                                                    Index node_ind_ofs,
+                                                    pr::ndarray<Index, 1>& node_list,
+                                                    imp_data_t& left_child_imp_data_list,
+                                                    pr::ndarray<Float, 1>& node_imp_dec_list,
+                                                    bool update_imp_dec_required,
+                                                    Index node_count,
+                                                    const bk::event_vector& deps = {});
+
+    static sycl::event best_split_single_pass_small(sycl::queue& queue,
+                                                    const context_t& ctx,
+                                                    const pr::ndarray<Bin, 2>& data,
+                                                    const pr::ndview<Float, 1>& response,
+                                                    const pr::ndarray<Index, 1>& tree_order,
+                                                    const pr::ndarray<Index, 1>& selected_ftr_list,
+                                                    const pr::ndarray<Index, 1>& bin_offset_list,
+                                                    const imp_data_t& imp_data_list,
+                                                    const node_group_view_t& node_group,
+                                                    node_list_t& level_node_list,
+                                                    imp_data_t& left_child_imp_data_list,
+                                                    pr::ndarray<Float, 1>& node_imp_dec_list,
+                                                    bool update_imp_dec_required,
+                                                    const bk::event_vector& deps = {});
+
+    static sycl::event random_split_single_pass(sycl::queue& queue,
```

Comment:
I vote for adding some comments at least at the beginning of the functions like this. I understand that there were no good commenting before in this file, but let's try to improve it.
This function has 17 arguments and cyclomatic complexity == 22 as CodeFactor [claims](https://www.codefactor.io/repository/github/oneapi-src/onedal/pull/2270). So, without any explanation it is extremely hard to understand what is going on here. And I think you'll agree with me about this statement.

An example of a comment can be found here: https://github.com/oneapi-src/oneDAL/blob/master/cpp/oneapi/dal/table/homogen.hpp